Libplanet/Blocks/PreEvaluationBlockHeader.cs

```diff
@@ -270,20 +296,109 @@ ImmutableArray<byte> preEvaluationHash
+                    "The block with the protocol version < 2 cannot be signed, because it lacks " +
+                    "its miner's public key so that others cannot verify its signature."
+                );
+            }
+            else if (!privateKey.PublicKey.Equals(PublicKey))
+            {
+                string m = "The given private key does not match to the miner's public key." +
+                    $"Block's public key: {PublicKey}\n" +
+                    $"Derived public key: {privateKey.PublicKey}\n";
+                throw new ArgumentException(m, nameof(privateKey));
+            }
+
+            byte[] msg = Codec.Encode(ToBencodex(stateRootHash));
+            byte[] sig = privateKey.Sign(msg);
+            return ImmutableArray.Create(sig);
+        }
+
+        /// <summary>
+        /// Verifies if the given block <paramref name="signature"/> is valid with the block
+        /// content and the specified <paramref name="stateRootHash"/>.
+        /// </summary>
+        /// <param name="signature">The block signature created using <see cref="MakeSignature"/>
+        /// method with the <paramref name="stateRootHash"/>.  This must be <c>null</c> for
+        /// blocks with earlier <seealso cref="ProtocolVersion"/>s than 2.</param>
+        /// <param name="stateRootHash">The state root hash included in the original message
+        /// of the signature.</param>
+        /// <returns><c>true</c> if the signature is valid.  <c>false</c> otherwise.</returns>
+        public bool VerifySignature(
+            ImmutableArray<byte>? signature,
+            HashDigest<SHA256> stateRootHash
+        )
+        {
+            if (PublicKey is { } pubKey && signature is { } sig)
+            {
+                byte[] msg = Codec.Encode(ToBencodex(stateRootHash));
+                return pubKey.Verify(msg, sig);
+            }
+            else if (PublicKey is null)
+            {
+                return signature is null;
+            }
+
+            return false;
+        }
 
         /// <summary>
         /// Derives a hash digest from the given pre-evaluation block header and
         /// <paramref name="stateRootHash"/>.
         /// </summary>
         /// <param name="stateRootHash">The state root hash.</param>
+        /// <param name="signature">The block signature created using <see cref="MakeSignature"/>
+        /// method with the <paramref name="stateRootHash"/>.  This must be <c>null</c> for
+        /// blocks with earlier <seealso cref="ProtocolVersion"/>s than 2.</param>
         /// <returns>A block hash.</returns>
         public BlockHash DeriveBlockHash(
-            in HashDigest<SHA256> stateRootHash
+            in HashDigest<SHA256> stateRootHash,
+            ImmutableArray<byte>? signature
         ) =>
-            BlockHash.DeriveFrom(Codec.Encode(ToBencodex(stateRootHash)));
+            BlockHash.DeriveFrom(Codec.Encode(ToBencodex(stateRootHash, signature)));
```

Comment:
I am not sure that the state root hash should be included here or not.
Could you explain the rationale?